frontend/src/metabase/visualizations/components/settings/ChartSettingColumnEditor.tsx

```diff
@@ -0,0 +1,204 @@
+/* eslint-disable react/prop-types */
+import React, { useMemo } from "react";
+import { t } from "ttag";
+import _ from "underscore";
+
+import Button from "metabase/core/components/Button";
+import { DatasetColumn } from "metabase-types/api";
+import { ConcreteField, Field } from "metabase-types/types/Query";
+import { isNotNull } from "metabase/core/utils/types";
+
+import StructuredQuery from "metabase-lib/queries/StructuredQuery";
+import NativeQuery from "metabase-lib/queries/NativeQuery";
+import Question from "metabase-lib/Question";
+import Dimension from "metabase-lib/Dimension";
+
+import {
+  TableHeaderContainer,
+  TableName,
+  FieldCheckbox,
+} from "./ChartSettingColumnEditor.styled";
+
+type ColumnSetting = {
+  name?: string;
+  fieldRef?: Field | null;
+  enabled: boolean;
+};
+
+interface ChartSettingColumnEditorProps {
+  value: ColumnSetting[];
+  onChange: (val: ColumnSetting[]) => void;
+  columns: DatasetColumn[];
+  question: Question;
+}
+
+const ChartSettingColumnEditor = ({
+  onChange,
+  value: columnSettings,
+  columns,
+  question,
+}: ChartSettingColumnEditorProps) => {
+  const fieldOptions = useMemo(() => {
```

Comment:
Do you see a path towards less complexity in this function?

Reply:
I could probably spread some of the complexity from here to the rest of the component, though it will probably be less efficient. Unfortunately, this needs handle Structured and Native queries quite differently.

Reply:
Is something like this viable/interesting as you look at things?

Done quick and dirty.
It's just reducing cyclomatic complexity, and unsurprisingly, levels of indentation.

```js
  const getFieldOptionsForStructuredQuery = (query: any, options: any) => ({
    name: query.table()?.display_name || t`Columns`,
    dimensions: options.dimensions.filter(isNotNull),
    fks: options.fks.map((fk: any) => ({
      name:
        fk.name ||
        (fk.field.target
          ? fk.field.target.table?.display_name
          : fk.field.displayName()),
      dimensions: fk.dimensions.filter(isNotNull),
    })),
  });

  const getFieldOptionsForNativeQuery = (query: any, columns: any) => ({
    name: query.table()?.display_name || t`Columns`,
    dimensions: columns
      .map((column: any) =>
        Dimension.parseMBQL(column.field_ref as ConcreteField),
      )
      .filter(isNotNull),
    fks: [],
  });

  const fieldOptions = useMemo(() => {
    const query = question && question.query();
    if (query instanceof StructuredQuery) {
      const options = query.fieldsOptions();
      return getFieldOptionsForStructuredQuery(query, options);
    } else if (query instanceof NativeQuery && columns) {
      return getFieldOptionsForNativeQuery(query, columns);
    } else {
      return {
        count: 0,
        dimensions: [],
        fks: [],
      };
    }
  }, [question, columns]);
```

Reply:
Oh! Yeah, can definitely split up the function a bit 👍

Reply:
I'd consider even more strongly breaking down this function as it's grown larger.

Reply:
🤦 You're right. I forgot to circle back to this